sdk/core/azure-core-experimental/src/main/java/com/azure/core/experimental/http/policy/ARMChallengeAuthenticationPolicy.java

```diff
@@ -0,0 +1,129 @@
+import java.util.ArrayList;
+import java.util.Base64;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.NoSuchElementException;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+/**
+ * The pipeline policy that applies a token credential to an HTTP request
+ * with "Bearer" scheme in ARM challenge based authentication scenarios
+ * .
+ * This is a temporary class to support ARM Challenge based authentication. It will
+ * move to azure-resource-manager package.
+ */
+public class ARMChallengeAuthenticationPolicy extends BearerTokenAuthenticationChallengePolicy {
+    private static final Pattern AUTHENTICATION_CHALLENGE_PATTERN =
+        Pattern.compile("(\\w+) ((?:\\w+=\".*?\"(?:, )?)+)(?:, )?");
+    private static final Pattern AUTHENTICATION_CHALLENGE_PARAMS_PATTERN =
+        Pattern.compile("(?:(\\w+)=\"([^\"\"]*)\")+");
+    private static final String CLAIMS_PARAMETER = "claims";
+    private final String[] scopes;
+    private final AzureEnvironment environment;
+    private static final String ARM_SCOPES_KEY = "ARMScopes";
+
+    /**
+     * Creates ARMChallengeAuthenticationPolicy.
+     *
+     * @param credential the token credential to authenticate the request
+     * @param environment the environment with endpoints for authentication
+     * @param scopes the scopes used in credential, using default scopes when empty
+     */
+    public ARMChallengeAuthenticationPolicy(TokenCredential credential,
+                                            AzureEnvironment environment, String... scopes) {
+        super(credential);
+        this.scopes = scopes;
+        this.environment = environment;
+    }
+
+
+    @Override
+    public Mono<Void> onBeforeRequest(HttpPipelineCallContext context) {
+        return Mono.defer(() -> {
+            TokenRequestContext trc;
+            String[] scopes = this.scopes;
+            if (scopes == null || scopes.length == 0) {
+                scopes = new String[1];
+                scopes[0] = ARMScopeHelper.getDefaultScopeFromRequest(
+                    context.getHttpRequest(), environment);
+            }
+            trc = new TokenRequestContext().addScopes(scopes);
+            context.setData(ARM_SCOPES_KEY, scopes);
+            return authorizeRequest(context, trc);
+        });
+    }
+
+    @Override
+    public Mono<Boolean> onChallenge(HttpPipelineCallContext context, HttpResponse response) {
+        return Mono.defer(() -> {
+            String authHeader = response.getHeaderValue(WWW_AUTHENTICATE);
+            if (response.getStatusCode() == 401 && authHeader != null) {
+                List<AuthenticationChallenge> challenges = parseChallenges(authHeader);
+                for (AuthenticationChallenge authenticationChallenge : challenges) {
+                    Map<String, String> extractedChallengeParams =
+                        parseChallengeParams(authenticationChallenge.getChallengeParameters());
```

Comment:
Is it possible for this to return null?

Reply:
it won't return null, an empty map currently.